### PR TITLE
Remove some unnecessary list and array allocations.

### PR DESCRIPTION
**Pull request by JonHanna**

### CODE REVIEW

**Comment by stakx** [comment]
> ```c#
> #if NET20 || NET35
>    .ToArray()
> #endif
> ```

You could perhaps do the same in:

* `Linq\JsonPath\FieldMultipleFilter.cs`, around line 42.
* `JsonValidatingReader.cs`, around line 720.

File: Src/Newtonsoft.Json/JsonValidatingReader.cs [hunk 1]
```diff
@@ -710,15 +710,14 @@ private void ValidateEndObject(JsonSchemaModel schema)
 
             Dictionary<string, bool> requiredProperties = _currentScope.RequiredProperties;
 
-            if (requiredProperties != null)
+            if (requiredProperties != null && requiredProperties.Values.Any(v => !v))
```
Comment (JonHanna) [inline Src/Newtonsoft.Json/JsonValidatingReader.cs]: This benefits the non-throwing branch at the cost of more work to the throwing, which is generally a win for most cases.

File: Src/Newtonsoft.Json/Serialization/DefaultContractResolver.cs [hunk 2]
```diff
@@ -645,16 +647,24 @@ private ConstructorInfo GetAttributeConstructor(Type objectType)
 
         private ConstructorInfo GetParameterizedConstructor(Type objectType)
         {
-            IList<ConstructorInfo> constructors = objectType.GetConstructors(BindingFlags.Public | BindingFlags.Instance).ToList();
-
-            if (constructors.Count == 1)
+            var constructors = objectType.GetConstructors(BindingFlags.Public | BindingFlags.Instance);
```
Comment (JamesNK) [inline Src/Newtonsoft.Json/Serialization/DefaultContractResolver.cs]: Specify type

Comment (JonHanna) [inline Src/Newtonsoft.Json/Serialization/DefaultContractResolver.cs]: Done

**Pull request merged**